Clean up longest_match variants and abstract match comparisons to compare258

nmoinvaz opened the pull request:
This PR changes cleans up all the `longest_match` variants by abstracting the memory comparison code into `compare258` functions. There is static versions uses for `longest_match` which is in `match_p.h`. And there is non-static versions for use in `deflate_quick`. By changing `deflate_quick` to use `functable.compare258` this allows for a future PR that will make `deflate_quick` cross-platform.

One major benefit to this PR is that the `longest_match` function can now use `compare258` with SSE 4.2 intrinsics where as before it was only used by `deflate_quick`. I have also added a 64-bit version of uanligned compare258 when SSE 4.2 is not supported.

The benefits of this PR are:

- Longest_match comparisons are abstracted out so that we can better support additional architectures
- All deflate algorithms can now take advantage of intrinsics in memory comparison
  - Previously only deflate_quick used compare258_sse4.
- Allows us to port deflate_quick to all other platforms
- Adds support for 64-bit unaligned comparison on platforms that don't support SSE4 or AVX2
- Unifies longest_match variants

For background see #496.
For previous PR see #544.

nmoinvaz committed:
Abstracted out architecture specific implementations of 258 byte comparison to compare258.

Dead2 commented:
This includes changes from at least one other PR, please clean up and rebase.

Dead2 commented:
This could also really benefit from being split up into at least two commits it looks like, reviewing this is difficult because there is so much to consider and look at all at once. Split it up into a commit for setting up the new infrastructure, then one or more for changing code over to using it (The match_p.h changes alone are more than enough for one commit really).

nmoinvaz commented:
@Dead2, thanks for the initial feedback. I can break it up into multiple commits like you suggested.

At another stage (after this PR), I want to change deflate_quick to use longest_match and then change compare258 to compare256. But that is a optimization down the road that requires its own performance tests.

nmoinvaz commented:
Rebased.

nmoinvaz committed:
Combine longest_match implementations and use compare258 functable stub.

nmoinvaz commented:
Ok it wasn't easy but I tried to break up the commit the best I can.

Dead2 commented on CMakeLists.txt:
```diff
@@ -525,6 +557,22 @@ if(BASEARCH_X86_FOUND)
+        }"
+        HAVE_SSE42CMPSTR_INTRIN
+    )
+    set(CMAKE_REQUIRED_FLAGS)
```
CMAKE_REQUIRED_FLAGS was not set to enable SSE4.2, so I suspect this check will fail.
Perhaps remove the reset of the flag above the test?

Dead2 commented on arch/x86/compare258_avx.c:
```diff
@@ -0,0 +1,60 @@
+        ymm_xor = _mm256_xor_si256(ymm_src0, ymm_src1); // identical bytes = 00
+        ymm_xor = _mm256_cmpeq_epi8(ymm_xor, ymm_zero); // non-identical bytes = 00, identical bytes = FF
+        int mask = _mm256_movemask_epi8(ymm_xor); 
+        if (mask != 0xFFFFFFFF) {
```
/arch/x86/compare258_avx.c:40:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]

mtl1979 replied:
0xFFFFFFFF is same as INT32_C(-1) but in hexadecimal it is more obvious why it needs to be that specific number.

Dead2 commented on CMakeLists.txt:
```diff
@@ -648,23 +680,30 @@ if(WITH_OPTIM)
             endif()
         endif()
+        if(HAVE_SSE42CMPSTR_INTRIN)
+            add_definitions(-DX86_SSE42_CMP_STR)
```
This test and define seems to be missing from configure.

nmoinvaz replied:
Fixed.

Dead2 commented:
Results on my Skylake-based linux machine:
1.34% slower with configure
6.43% slower with cmake

cmake build was already slower, but with this, the default cmake build is 15.0% slower than the default configure build. Hoping this gets fixed with the -avx2 etc flags getting applied to only the correct files.

nmoinvaz commented:
More discouraging results. Are those the results for build time, compile time, or run time?

Dead2 commented:
Those are the speeds of the finished minigzip

nmoinvaz commented:
I have added the check to the configure script, fixed the check in the cmake script, tried to resolve comments, and also modified compare258 so first 2 bytes are checked at the end instead of the beginning of the functions for 32, 64, avx2, and sse4.

mtl1979 commented on functable.c:
```diff
@@ -157,3 +195,51 @@ ZLIB_INTERNAL uint32_t crc32_stub(uint32_t crc, const unsigned char *buf, uint64
+    if (x86_cpu_has_avx2)
+        functable.compare258 = &compare258_unaligned_avx2;
+#  endif
+#  ifdef X86_SSE42_CMP_STR
```
AVX2 version should come after SSE4.2 version as otherwise it will never be used.

nmoinvaz replied:
In my tests SSE4.2 was a bit faster.

mtl1979 replied:
That only means we have to try to optimize it further...

mtl1979 replied:
```
static inline int32_t compare258_unaligned_avx2_static(const unsigned char *src0, const unsigned char *src1) {
    const unsigned char *src0start = src0;
    const unsigned char *src0end = src0 + 256;
 
    do {
        __m256i ymm_src0, ymm_src1, ymm_cmp;
        ymm_src0 = _mm256_loadu_si256((__m256i*)src0);
        ymm_src1 = _mm256_loadu_si256((__m256i*)src1);
        ymm_cmp = _mm256_cmpeq_epi8(ymm_src0, ymm_src1); // non-identical bytes = 00, identical bytes = FF
        int mask = _mm256_movemask_epi8(ymm_cmp); 
        if ((unsigned int)mask != 0xFFFFFFFF) {
            int match_byte = __builtin_ctz(~mask); // Invert bits so identical = 0
            return (int32_t)(src0 - src0start + match_byte);
        }

        src0 += 32, src1 += 32;
    } while (src0 < src0end);

    if (*(uint16_t *)src0 == *(uint16_t *)src1)
        src0 += 2, src1 += 2;
    else if (*src0 == *src1)
        src0 += 1, src1 += 1;

    return (int32_t)(src0 - src0start);
}
```

This version doesn't use xor at all... basically it compares the operands directly... this works because we assume all ones is the best case scenario instead of all zeroes.

nmoinvaz replied:
Now AVX2 matches or beats SSE4.2

mtl1979 replied:
That's good... It just took little extra effort due to AVX2 not having cmpne... Reversing the tests and using GPR-version of NOT was the best solution... AVX512 has cmpne, so it will improve performance even more when more machines have it...

nmoinvaz replied:
It believe it should be: `__builtin_ctz(~mask) / 8`.

mtl1979 replied:
Nah... the _mm256_movemask_epi8() already strips the extra 7 bits...

mtl1979 commented on functable.c:
```diff
@@ -157,3 +195,51 @@ ZLIB_INTERNAL uint32_t crc32_stub(uint32_t crc, const unsigned char *buf, uint64
+    if (x86_cpu_has_avx2)
+        functable.longest_match = &longest_match_unaligned_avx2;
+#  endif
+#  ifdef X86_SSE42_CMP_STR
```
Same as above, AVX2 version should come after SSE4.2 version.

nmoinvaz replied:
Now AVX2 matches or beats SSE4.2

nmoinvaz commented:
[Here](https://gist.github.com/nmoinvaz/b38b30fa0742d3749dba082222f08892) are my performance tests using @neurolabusc's pigz-bench-python.

I haven't incorporated the AVX2 changes yet.

nmoinvaz commented:
I have simplified the original PR and removed `longest_match` from `functable`. I unified `longest_match` implementations into one which just uses `functable.compare258`. And I have updated my performance results [here](https://gist.github.com/nmoinvaz/b38b30fa0742d3749dba082222f08892).

nmoinvaz commented:
Rebased.

nmoinvaz commented:
Rebased

nmoinvaz commented:
Here are two performance tests I ran against this PR.

![compare258-x3](https://user-images.githubusercontent.com/1364432/80895282-f82efa00-8c97-11ea-9efb-1516a0e7c996.png)
![compare258-x2](https://user-images.githubusercontent.com/1364432/80895283-f8c79080-8c97-11ea-958d-b05dd3aae718.png)

nmoinvaz commented:
Once this is merged I can make an additional PR that should finally resolve #205.

Dead2 commented:
Rebase needed :)

Dead2 commented on compare258.c:
```diff
@@ -0,0 +1,122 @@
+#include "fallback_builtins.h"
+
+/* ALIGNED, byte comparison */
+int32_t compare258(const unsigned char *src0, const unsigned char *src1) {
```
_c postfix for the function name(s)?

nmoinvaz replied:
Done

nmoinvaz commented:
Rebase complete.

Dead2 commented:
But compressed size is identical, so that is good.
But this is unfortunately quite a bit slower than HEAD is. About 2.34% slower overall.

```
# HEAD
   text    data     bss     dec     hex filename
  86655    1216      32   87903   1575f libz-ng.so.1

 Runs: 30
 Levels: 1-8
 Skipworst: True (15)

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.227/0.232/0.235       0.129/0.133/0.135  20804915
 2      38.062%    0.408/0.417/0.421       0.130/0.135/0.137  15964205
 3      37.021%    0.516/0.525/0.530       0.128/0.131/0.134  15527698
 4      35.832%    0.587/0.599/0.606       0.124/0.129/0.132  15028898
 5      35.719%    0.642/0.653/0.659       0.122/0.128/0.131  14981562
 6      35.651%    0.788/0.800/0.810       0.125/0.129/0.132  14953264
 7      35.030%    1.442/1.457/1.467       0.124/0.128/0.131  14692564
 8      34.995%    2.232/2.249/2.269       0.124/0.127/0.130  14677992

 avg1   37.739%                0.867                   0.130
 tot                         103.996                  15.599
```

```
# PR/550 only first commit
   text    data     bss     dec     hex filename
  88302    1248      32   89582   15dee libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.228/0.231/0.235       0.129/0.132/0.134  20804915
 2      38.062%    0.412/0.420/0.424       0.130/0.134/0.137  15964205
 3      37.021%    0.516/0.524/0.530       0.129/0.132/0.134  15527698
 4      35.832%    0.595/0.601/0.607       0.125/0.130/0.132  15028898
 5      35.719%    0.649/0.659/0.664       0.124/0.129/0.131  14981562
 6      35.651%    0.791/0.802/0.814       0.125/0.129/0.131  14953264
 7      35.030%    1.455/1.467/1.479       0.125/0.129/0.131  14692564
 8      34.995%    2.243/2.261/2.277       0.124/0.128/0.130  14677992

 avg1   37.739%                0.871                   0.130
 tot                         104.472                  15.648

```

```
# PR/550
   text    data     bss     dec     hex filename
  88350    1248      32   89630   15e1e libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.227/0.232/0.235       0.126/0.133/0.136  20804915
 2      38.062%    0.435/0.441/0.446       0.129/0.135/0.137  15964205
 3      37.021%    0.535/0.548/0.553       0.125/0.131/0.133  15527698
 4      35.832%    0.597/0.608/0.615       0.126/0.129/0.131  15028898
 5      35.719%    0.646/0.660/0.668       0.124/0.128/0.130  14981562
 6      35.651%    0.786/0.801/0.808       0.119/0.127/0.130  14953264
 7      35.030%    1.486/1.504/1.519       0.123/0.127/0.130  14692564
 8      34.995%    2.291/2.306/2.323       0.124/0.128/0.131  14677992

 avg1   37.739%                0.887                   0.130
 tot                         106.487                  15.567
```

Dead2 commented:
Please try to separate out logical changes to different commits. For example compare258_avx2 could have easily been a separate commit added at the end, or even a separate PR.
This makes it easier to track down what commit is doing something weird.

Some of the slowdown comes from the avx2 implementation it seems.
This is the first commit without avx2:
```
# PR/550 only first commit without compare258_avx2
   text    data     bss     dec     hex filename
  88182    1240      32   89454   15d6e libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.225/0.231/0.234       0.127/0.132/0.135  20804915
 2      38.062%    0.418/0.423/0.426       0.134/0.137/0.138  15964205
 3      37.021%    0.518/0.525/0.531       0.128/0.132/0.135  15527698
 4      35.832%    0.594/0.603/0.608       0.124/0.129/0.132  15028898
 5      35.719%    0.648/0.655/0.665       0.126/0.128/0.132  14981562
 6      35.651%    0.795/0.804/0.814       0.127/0.129/0.132  14953264
 7      35.030%    1.440/1.452/1.462       0.123/0.128/0.130  14692564
 8      34.995%    2.224/2.250/2.268       0.123/0.128/0.131  14677992

 avg1   37.739%                0.868                   0.131
 tot                         104.122                  15.660
```

```
# PR/550 without compare258_avx2
   text    data     bss     dec     hex filename
  88230    1240      32   89502   15d9e libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.224/0.231/0.234       0.129/0.133/0.135  20804915
 2      38.062%    0.435/0.439/0.442       0.131/0.135/0.139  15964205
 3      37.021%    0.542/0.548/0.554       0.126/0.131/0.133  15527698
 4      35.832%    0.595/0.602/0.608       0.122/0.129/0.133  15028898
 5      35.719%    0.647/0.656/0.662       0.123/0.129/0.132  14981562
 6      35.651%    0.788/0.804/0.816       0.122/0.128/0.132  14953264
 7      35.030%    1.484/1.505/1.524       0.121/0.128/0.131  14692564
 8      34.995%    2.288/2.303/2.317       0.125/0.128/0.131  14677992

 avg1   37.739%                0.886                   0.130
 tot                         106.313                  15.609
```
That is all the time I have right now for manually experimenting with this.

nmoinvaz commented:
It is strange that my performance tests give a huge increase in compression speed (at least according to those graphs), but in your tests it decreases compression speed. I wonder if it has something to do with initial dataset. What is the dataset you are using?

I am using this set of corpus which is part of `pigz-bench-python` by @neurolabusc:

https://github.com/MiloszKrajewski/SilesiaCorpus
https://github.com/neurolabusc/zlib-bench.git

nmoinvaz committed:
Added support for AVX2 intrinsics to compare258.

nmoinvaz committed:
Modified longest_match variants to use associated compare258 functions.

nmoinvaz committed:
Removed fallback for __builtin_ctzl since it is no longer used.

nmoinvaz committed:
Modified deflate_quick to use compare258_unaligned_sse4.

nmoinvaz committed:
Changed deflate_quick to use compare258 functable stub.

nmoinvaz commented:
@Dead2 I have split the PR into more commits for you.

Dead2 commented:
My test dataset is a 40MiB .tar file that contains a mix of various long and short text files (text,html,xml,ini,etc) of some in different languages, mbox files, bitmap images, elf executables/libraries, some already compressed files and various other binary files.

In an effort to make a realistic mix, my dataset contains private data.

What generation cpu do you test with? The one I use currently is an i9-9900K (with several tweaks to avoid turbo, throttling, interrupts, etc). Operating system can also cause differences in some cases. Another reason for differences where there should be none is how the compiler lays things out in memory. Cache locality and cache aliasing are examples of things that can have a great effect on the performance of code that is completely unrelated to the actual code change.
You might have noticed in my deflate compression benchmarks, often the inflate speed will also change (and is repeatable), that is often because of memory layout (+ the normal variance). Usually there is not a huge change, but sometimes you get a golden or a absolutely rubbish layout.

Thanks for splitting this up like this, this is really good. I'll try get some benchmarks done.

nmoinvaz commented:
Ok I am using Windows 10 with Visual Studio 2019. I'm not sure how I would turn off some of those settings to enable more consistent results. What OS are you using?

![image](https://user-images.githubusercontent.com/1364432/81327941-e8753400-9050-11ea-91cf-fabbbb73209e.png)

Dead2 commented:
```
# HEAD
   text    data     bss     dec     hex filename
  86655    1216      32   87903   1575f libz-ng.so.1

 Runs: 30
 Levels: 1-8
 Skipworst: True (15)

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.227/0.232/0.235       0.129/0.133/0.135  20804915
 2      38.062%    0.408/0.417/0.421       0.130/0.135/0.137  15964205
 3      37.021%    0.516/0.525/0.530       0.128/0.131/0.134  15527698
 4      35.832%    0.587/0.599/0.606       0.124/0.129/0.132  15028898
 5      35.719%    0.642/0.653/0.659       0.122/0.128/0.131  14981562
 6      35.651%    0.788/0.800/0.810       0.125/0.129/0.132  14953264
 7      35.030%    1.442/1.457/1.467       0.124/0.128/0.131  14692564
 8      34.995%    2.232/2.249/2.269       0.124/0.127/0.130  14677992

 avg1   37.739%                0.867                   0.130
 tot                         103.996                  15.599

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.228/0.232/0.236       0.126/0.132/0.136  20804915
 2      38.062%    0.406/0.420/0.425       0.130/0.134/0.137  15964205
 3      37.021%    0.516/0.525/0.529       0.129/0.132/0.135  15527698
 4      35.832%    0.590/0.600/0.607       0.124/0.128/0.131  15028898
 5      35.719%    0.638/0.654/0.660       0.124/0.128/0.132  14981562
 6      35.651%    0.788/0.800/0.809       0.124/0.128/0.130  14953264
 7      35.030%    1.440/1.451/1.463       0.125/0.129/0.131  14692564
 8      34.995%    2.223/2.246/2.268       0.120/0.128/0.130  14677992

 avg1   37.739%                0.866                   0.130
 tot                         103.913                  15.568
```
```
# PR/550/1 3d704c5b692404abbfd18687682d5aa4ca83d555
   text    data     bss     dec     hex filename
  87956    1240      32   89228   15c8c libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.228/0.230/0.233       0.130/0.133/0.136  20804915
 2      38.062%    0.412/0.421/0.426       0.134/0.137/0.140  15964205
 3      37.021%    0.520/0.524/0.528       0.128/0.132/0.136  15527698
 4      35.832%    0.591/0.601/0.609       0.128/0.130/0.132  15028898
 5      35.719%    0.651/0.657/0.662       0.128/0.131/0.133  14981562
 6      35.651%    0.788/0.800/0.807       0.125/0.129/0.132  14953264
 7      35.030%    1.433/1.458/1.481       0.123/0.129/0.132  14692564
 8      34.995%    2.236/2.250/2.264       0.123/0.128/0.131  14677992

 avg1   37.739%                0.868                   0.131
 tot                         104.126                  15.744
```
```
# PR/550/2 65f8cd8e28f82d158c03613be5c3976cfb8ac7e9
   text    data     bss     dec     hex filename
  87883    1248      32   89163   15c4b libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.225/0.229/0.231       0.129/0.133/0.135  20804915
 2      38.062%    0.415/0.421/0.426       0.131/0.136/0.138  15964205
 3      37.021%    0.521/0.526/0.532       0.131/0.134/0.136  15527698
 4      35.832%    0.593/0.600/0.606       0.121/0.130/0.132  15028898
 5      35.719%    0.651/0.656/0.662       0.123/0.129/0.132  14981562
 6      35.651%    0.793/0.804/0.811       0.122/0.128/0.131  14953264
 7      35.030%    1.438/1.455/1.472       0.123/0.127/0.129  14692564
 8      34.995%    2.230/2.251/2.268       0.122/0.129/0.131  14677992

 avg1   37.739%                0.868                   0.131
 tot                         104.119                  15.675
```
```
# PR/550/3 521fc6ea008e973039354d5a83f91dd3723a0d36
   text    data     bss     dec     hex filename
  87811    1256      32   89099   15c0b libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.226/0.230/0.233       0.129/0.132/0.135  20804915
 2      38.062%    0.428/0.434/0.438       0.132/0.136/0.137  15964205
 3      37.021%    0.539/0.548/0.554       0.126/0.132/0.134  15527698
 4      35.832%    0.599/0.612/0.621       0.122/0.129/0.132  15028898
 5      35.719%    0.656/0.667/0.674       0.122/0.130/0.133  14981562
 6      35.651%    0.804/0.816/0.824       0.119/0.129/0.132  14953264
 7      35.030%    1.486/1.496/1.508       0.123/0.127/0.130  14692564
 8      34.995%    2.305/2.329/2.344       0.122/0.127/0.131  14677992

 avg1   37.739%                0.892                   0.130
 tot                         106.983                  15.637
```
```
# PR/550/4 4164726ee0b07672ccc16da9eca21323a7f4cbce
   text    data     bss     dec     hex filename
  87811    1256      32   89099   15c0b libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.226/0.231/0.233       0.129/0.133/0.135  20804915
 2      38.062%    0.430/0.436/0.440       0.131/0.135/0.137  15964205
 3      37.021%    0.541/0.549/0.555       0.124/0.131/0.133  15527698
 4      35.832%    0.609/0.617/0.623       0.125/0.128/0.131  15028898
 5      35.719%    0.664/0.671/0.676       0.125/0.129/0.130  14981562
 6      35.651%    0.801/0.816/0.827       0.124/0.128/0.131  14953264
 7      35.030%    1.489/1.506/1.521       0.123/0.126/0.129  14692564
 8      34.995%    2.301/2.325/2.338       0.124/0.128/0.130  14677992

 avg1   37.739%                0.894                   0.130
 tot                         107.272                  15.575
```
```
# PR/550/5 8c4a9eaea9f5511cb66a33a7a95ee160c9f2bffd
   text    data     bss     dec     hex filename
  87798    1248      32   89078   15bf6 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.225/0.230/0.234       0.129/0.133/0.135  20804915
 2      38.062%    0.426/0.435/0.440       0.132/0.136/0.139  15964205
 3      37.021%    0.545/0.550/0.554       0.129/0.132/0.135  15527698
 4      35.832%    0.607/0.616/0.622       0.122/0.129/0.132  15028898
 5      35.719%    0.666/0.675/0.680       0.125/0.129/0.132  14981562
 6      35.651%    0.797/0.815/0.823       0.124/0.128/0.132  14953264
 7      35.030%    1.485/1.501/1.513       0.121/0.127/0.130  14692564
 8      34.995%    2.313/2.330/2.346       0.126/0.128/0.130  14677992

 avg1   37.739%                0.894                   0.130
 tot                         107.281                  15.648
```
```
# PR/550/6 f26b46ada14db15b821e72fc9b31407c42e2ee46
   text    data     bss     dec     hex filename
  87918    1240      32   89190   15c66 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.227/0.230/0.233       0.125/0.132/0.134  20804915
 2      38.062%    0.431/0.439/0.443       0.131/0.134/0.137  15964205
 3      37.021%    0.539/0.546/0.553       0.127/0.132/0.135  15527698
 4      35.832%    0.593/0.602/0.608       0.123/0.129/0.132  15028898
 5      35.719%    0.645/0.656/0.667       0.126/0.130/0.133  14981562
 6      35.651%    0.790/0.801/0.814       0.125/0.130/0.132  14953264
 7      35.030%    1.482/1.504/1.519       0.125/0.128/0.131  14692564
 8      34.995%    2.281/2.309/2.339       0.118/0.128/0.131  14677992

 avg1   37.739%                0.886                   0.130
 tot                         106.314                  15.641
```
```
# PR/550/7 2e12cd914e8bebab3f2144d2f5874ed2c946a207
   text    data     bss     dec     hex filename
  88350    1248      32   89630   15e1e libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.226/0.231/0.234       0.128/0.132/0.136  20804915
 2      38.062%    0.433/0.441/0.445       0.129/0.134/0.137  15964205
 3      37.021%    0.539/0.547/0.550       0.129/0.132/0.135  15527698
 4      35.832%    0.599/0.606/0.609       0.126/0.129/0.131  15028898
 5      35.719%    0.651/0.657/0.663       0.125/0.129/0.131  14981562
 6      35.651%    0.795/0.803/0.812       0.122/0.128/0.131  14953264
 7      35.030%    1.483/1.501/1.514       0.122/0.127/0.130  14692564
 8      34.995%    2.289/2.314/2.336       0.125/0.129/0.130  14677992

 avg1   37.739%                0.887                   0.130
 tot                         106.490                  15.620

```

nmoinvaz commented:
How does the first commit slow things down? That doesn't seem right, because nothing is hooked up at that point. I did notice `Skipworst: True (15)` is only set for `HEAD`.

Dead2 commented:
skipworst is set for them all, I just don't copy-paste the settings-info in for all of them.

That difference is quite small and is probably caused by the compiled code location changing, this can affect both of cache locality (closer is better) and cache aliasing (cacheline collisions are bad, but almost impossible to predict unless you specifically code for it).

nmoinvaz commented:
I'm not sure how to overcome that and I'm not sure what to do with this PR. Are you using configure or CMake?

Dead2 commented:
Patch #3 is the one that has a big increase in total runtime, going from ~104s to ~107s, so that is where there should be a potential for improvement.

mtl1979 commented:
521fc6ea008e973039354d5a83f91dd3723a0d36 slows down because it uses extern, when it could inline the compare functions... This might mean some of the versions might benefit from using own copy of the compare functions.

Basically only when the vector versions are required/preferred, we need to go through functable.

nmoinvaz commented:
@mtl1979 that is what I originally had with this PR. I had longest_match in functable..

mtl1979 commented:
@nmoinvaz Sometimes we need to go back and forth to find the best compromise... I was only commenting based on single commit.

I see it as compare258 being called so many times that going through even extern slows it down too much... It is possible to have multiple copies of same function in executable if it is declared as static inline in a header file. I think we have done it somewhere before already...

nmoinvaz committed:
Converted compare258 to static and convert longest_match to template.

nmoinvaz committed:
Split compare258 into compare256 for longest_match and compare258 for deflate_quick.

nmoinvaz commented:
@Dead2 I have made a few more commits to try some things. Can you please do some performance tests against each of them? Thank you.

Dead2 commented:
```
# PR/550/8 b19b4675251a3f66ddd3ae1bdb12d8337a413f06
   text    data     bss     dec     hex filename
  88462    1248      32   89742   15e8e libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.224/0.229/0.233       0.127/0.133/0.136  20804915
 2      38.062%    0.441/0.449/0.453       0.130/0.134/0.137  15964205
 3      37.021%    0.544/0.554/0.559       0.128/0.132/0.135  15527698
 4      35.832%    0.600/0.611/0.619       0.125/0.130/0.132  15028898
 5      35.719%    0.659/0.668/0.675       0.126/0.129/0.132  14981562
 6      35.651%    0.801/0.817/0.829       0.126/0.130/0.132  14953264
 7      35.030%    1.491/1.518/1.538       0.121/0.127/0.131  14692564
 8      34.995%    2.309/2.330/2.342       0.123/0.127/0.130  14677992

 avg1   37.739%                0.897                   0.130
 tot                         107.631                  15.625
```

```
# PR/550/9 1b4f860d0ff793a22e295c4bc1b4dddd7efb7cf0
   text    data     bss     dec     hex filename
  90717    1280      32   92029   1677d libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.226/0.229/0.231       0.125/0.132/0.135  20804915
 2      38.062%    0.431/0.440/0.443       0.129/0.136/0.138  15964205
 3      37.021%    0.537/0.545/0.554       0.127/0.131/0.134  15527698
 4      35.832%    0.596/0.607/0.615       0.127/0.130/0.132  15028898
 5      35.719%    0.648/0.660/0.669       0.125/0.129/0.132  14981562
 6      35.651%    0.791/0.809/0.821       0.126/0.129/0.131  14953264
 7      35.030%    1.464/1.477/1.497       0.122/0.128/0.131  14692564
 8      34.995%    2.262/2.280/2.294       0.121/0.127/0.130  14677992

 avg1   37.739%                0.881                   0.130
 tot                         105.727                  15.647
```

```
# PR/550/10 e7bee25fed41dc66b6afdbcff61ff3143c548afc
   text    data     bss     dec     hex filename
  90493    1280      32   91805   1669d libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.227/0.229/0.233       0.127/0.132/0.134  20804915
 2      38.071%    0.430/0.435/0.440       0.130/0.134/0.137  15968288
 3      37.028%    0.523/0.536/0.540       0.128/0.132/0.134  15530861
 4      35.838%    0.596/0.601/0.608       0.124/0.129/0.132  15031690
 5      35.678%    0.651/0.657/0.662       0.124/0.129/0.131  14964370
 6      35.614%    0.794/0.804/0.811       0.126/0.129/0.132  14937479
 7      35.041%    1.449/1.459/1.478       0.123/0.127/0.130  14697174
 8      35.007%    2.244/2.261/2.272       0.126/0.129/0.132  14682859

 avg1   37.735%                0.873                   0.130
 tot                         104.747                  15.615

```

```
# PR/550/11 eebdc54cbf726564b4eccff865a5e0a532dd4a47
   text    data     bss     dec     hex filename
  90174    1280      32   91486   1655e libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      60.508%    0.307/0.313/0.315       0.143/0.148/0.151  25378972
 2      38.071%    0.429/0.436/0.440       0.132/0.135/0.138  15968288
 3      37.028%    0.533/0.540/0.545       0.127/0.132/0.135  15530861
 4      35.838%    0.591/0.600/0.605       0.127/0.130/0.132  15031690
 5      35.678%    0.648/0.658/0.664       0.125/0.129/0.131  14964370
 6      35.614%    0.788/0.798/0.808       0.125/0.129/0.130  14937479
 7      35.041%    1.441/1.464/1.487       0.121/0.127/0.131  14697174
 8      35.007%    2.243/2.263/2.277       0.121/0.127/0.130  14682859

 avg1   39.098%                0.884                   0.132
 tot                         106.082                  15.858
```

Note that the last two change the compression ratio.
#10 looks like it makes a random-ish impact on compression
#11 severely degrades deflate_quick compression ratio

#9 improves speed without any ill-effects, not quite back to HEAD's performance, but it does remove a lot of the performance penalty.

nmoinvaz commented:
I think some of the instances where things were a bit weird were due to how I was calling `compare256`. I have fix that and performance results should improve @Dead2.

I am open to any further suggestions on how I can speed things up. @mtl1979

Dead2 commented:
It seems that this fails regression testing in some cases.

Will run some benchmarking after dinner.

nmoinvaz commented:
Good catch. Should be fixed now.

Dead2 commented:
This is looking a lot better.

```
# PR/550/8 32ab46659bff33c15c23d7e48b2a083d581e6636
   text    data     bss     dec     hex filename
  90597    1280      32   91909   16705 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.221/0.229/0.233       0.128/0.133/0.136  20804915
 2      38.062%    0.433/0.441/0.445       0.131/0.136/0.138  15964205
 3      37.021%    0.532/0.543/0.548       0.128/0.133/0.135  15527698
 4      35.832%    0.596/0.606/0.614       0.127/0.131/0.132  15028898
 5      35.719%    0.647/0.659/0.670       0.128/0.131/0.133  14981562
 6      35.651%    0.789/0.803/0.813       0.123/0.129/0.132  14953264
 7      35.030%    1.450/1.470/1.485       0.121/0.128/0.130  14692564
 8      34.995%    2.238/2.257/2.280       0.123/0.128/0.130  14677992

 avg1   37.739%                0.876                   0.131
 tot                         105.118                  15.754

```

```
# PR/550/9 894feee471fc1ec8a2771461d2ab2228b7554ee6
   text    data     bss     dec     hex filename
  90573    1280      32   91885   166ed libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.224/0.230/0.234       0.129/0.132/0.136  20804915
 2      38.062%    0.431/0.435/0.438       0.132/0.135/0.138  15964205
 3      37.021%    0.530/0.537/0.543       0.130/0.134/0.136  15527698
 4      35.832%    0.590/0.600/0.604       0.122/0.128/0.132  15028898
 5      35.719%    0.646/0.657/0.666       0.126/0.129/0.132  14981562
 6      35.651%    0.779/0.795/0.808       0.124/0.129/0.132  14953264
 7      35.030%    1.440/1.460/1.477       0.122/0.128/0.130  14692564
 8      34.995%    2.236/2.259/2.280       0.125/0.129/0.131  14677992

 avg1   37.739%                0.871                   0.131
 tot                         104.577                  15.682

```

```
# PR/550/10 de27ee8656afb9ae0676a516bec980af7bcd1a04
   text    data     bss     dec     hex filename
  90228    1280      32   91540   16594 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.226/0.231/0.233       0.131/0.135/0.136  20804915
 2      38.062%    0.432/0.437/0.443       0.131/0.135/0.138  15964205
 3      37.021%    0.530/0.539/0.543       0.126/0.133/0.135  15527698
 4      35.832%    0.593/0.601/0.607       0.124/0.130/0.132  15028898
 5      35.719%    0.640/0.656/0.665       0.123/0.130/0.133  14981562
 6      35.651%    0.779/0.798/0.812       0.127/0.131/0.133  14953264
 7      35.030%    1.446/1.465/1.476       0.119/0.128/0.131  14692564
 8      34.995%    2.242/2.254/2.270       0.126/0.129/0.132  14677992

 avg1   37.739%                0.873                   0.131
 tot                         104.724                  15.752
```

nmoinvaz commented:
Interesting, it looks like the last commit hurt `deflate_quick` performance a bit. I can remove it, it is not essential.

Dead2 commented:
It looks like this introduces some new warnings. Try running `./configure --warn && make`

nmoinvaz commented:
Ok those warnings are resolved now it looks like.

Dead2 commented:
~0.76% slower than HEAD, that is acceptable but not optimal.
I have only tested x86_64 with configure.
I do wonder what the impact is on other machines, ARM especially.

Do the benefits of the changes warrant the small performance drop?
I'd like others to weigh in on this too, and also review the code changes if possible.. @mtl1979 @sebpop @mscastanho & others?

mtl1979 commented on arch/x86/compare258_avx.c:
```diff
@@ -0,0 +1,73 @@
+#  include <nmmintrin.h>
+#endif
+
+/* UNALIGNED_OK, AVX2 instrinic comparison */
```
Spelling mistake, should be "intrinsic"

nmoinvaz replied:
Fixed.

mtl1979 commented on arch/x86/compare258_sse.c:
```diff
@@ -0,0 +1,81 @@
+#  include <nmmintrin.h>
+#endif
+
+/* UNALIGNED_OK, SSE4.2 instrinic comparison */
```
Spelling mistake, should be "intrinsic"

nmoinvaz replied:
Fixed.

mtl1979 commented on arch/x86/compare258_sse.c:
```diff
@@ -0,0 +1,81 @@
+    if (*src0 != *src1)
+        return 2;
+    if (src0[1] != src1[1])
+        return 3;
```
8-bit compare is really slow, this should be changed to 16-bit compare like with first two bytes.

nmoinvaz replied:
Ok I have made the change.

mtl1979 commented on arch/x86/compare258_avx.c:
```diff
@@ -0,0 +1,73 @@
+    if (*src0 != *src1)
+        return 2;
+    if (src0[1] != src1[1])
+        return 3;
```
8-bit compare is really slow, this should be changed to 16-bit compare like with first two bytes.

nmoinvaz replied:
Done.

mtl1979 commented on compare258.c:
```diff
@@ -0,0 +1,204 @@
+        return 2;
+    src0 += 1, src1 += 1;
+    if (*src0 != *src1)
+        return 3;
```
This makes no sense at all... 258 is multiple of many numbers and 258-compare can be implemented more efficiently without calling 256-byte compare.

nmoinvaz replied:
It depends on `MIN_MATCH`. `longest_match` will use COMPARE256 if MIN_MATCH > 1 and COMPARE258 if MIN_MATCH <= 1

nmoinvaz replied:
It is 256 because the initial match check is at least 2 bytes (uint16_t)

Dead2 replied:
Hmm, is MIN_MATCH lower than 3 even possible/allowed/sensible? I thought 3 was the minimum, but that you could increase this to avoid having short matches.

mtl1979 replied:
258 is multiple of 6, 256 is multiple of 4 and 8... This means it is more efficient to use 6 checks per iteration when comparing 258 bytes, but 4 checks per iteration when comparing 256 bytes.

Only few processors have enough general-purpose registers to do 8 checks per iteration efficiently.

mtl1979 replied:
@Dead2 It is just to make sure the initial 2 bytes match before going to the optimized version that compares 256 bytes instead of 258 bytes... It's inefficient to use intrinsics for the 2 bytes.

Dead2 replied:
Yes, I just don't understand @nmoinvaz comment about MIN_MATCH <= 1. That doesn't make sense to me.

mtl1979 replied:
@Dead2 To me neither... But that's why we ask clarifications... Some times we write something we don't mean to due to distractions ;)

nmoinvaz committed:
Remove additional byte comparison check in compare258 SSE4 and AVX2 functions.

nmoinvaz commented:
@Dead2, I have added an additional commit that should be checked for performance on its own. 
I rebased for misspelling that @mtl1979 brought to my attention.

nmoinvaz commented:
The benefits of this PR are:

* Longest_match comparisons are abstracted out so that we can better support additional architectures
* All deflate algorithms can now take advantage of intrinsics in memory comparison
  * Previously only deflate_quick used compare258_sse4.
* Allows us to port deflate_quick to all other platforms
* Adds support for 64-bit unaligned comparison on platforms that don't support SSE4 or AVX2
* Unifies longest_match variants

Dead2 commented:
```
# PR/550/10 cf221c6a5bb73a93fa6134319b05f22f00849933
   text    data     bss     dec     hex filename
  90525    1280      32   91837   166bd libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.225/0.230/0.232       0.126/0.133/0.136  20804915
 2      38.062%    0.431/0.436/0.440       0.131/0.135/0.137  15964205
 3      37.021%    0.527/0.534/0.540       0.129/0.132/0.133  15527698
 4      35.832%    0.599/0.605/0.610       0.126/0.131/0.133  15028898
 5      35.719%    0.641/0.654/0.660       0.125/0.128/0.131  14981562
 6      35.651%    0.782/0.797/0.808       0.124/0.129/0.132  14953264
 7      35.030%    1.446/1.459/1.473       0.123/0.127/0.130  14692564
 8      34.995%    2.233/2.245/2.265       0.121/0.128/0.131  14677992

 avg1   37.739%                0.870                   0.130
 tot                         104.391                  15.642
```
So about 0.4% slower than HEAD, that is indeed very good.


I wanted to do another test with the "With deflate_quick using compare256.." commit on top of this, this is what that would look like. The difference between this and the above is very low, and it is very hard to say whether this is a real difference or just normal variance. If it cleans up code, it might well be worth doing. deflate_quick is still faster than it is in HEAD.
```
# PR/550/11 "With deflate_quick using compare256.."
   text    data     bss     dec     hex filename
  90228    1280      32   91540   16594 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.227/0.231/0.233       0.128/0.133/0.135  20804915
 2      38.062%    0.428/0.436/0.440       0.132/0.136/0.138  15964205
 3      37.021%    0.528/0.534/0.537       0.125/0.132/0.135  15527698
 4      35.832%    0.595/0.600/0.606       0.127/0.131/0.134  15028898
 5      35.719%    0.635/0.651/0.659       0.126/0.130/0.132  14981562
 6      35.651%    0.782/0.797/0.810       0.124/0.130/0.132  14953264
 7      35.030%    1.441/1.458/1.469       0.124/0.128/0.130  14692564
 8      34.995%    2.233/2.260/2.281       0.126/0.129/0.130  14677992

 avg1   37.739%                0.871                   0.131
 tot                         104.493                  15.727
```

nmoinvaz commented:
I would say that “use compare256” commit isn’t really necessary. At the time I thought I could do that and get rid of compare258 but that turned out not to be the case because I still use it in longest_match when MIN_MATCH is 1.

nmoinvaz commented:
Nice performance results! Thanks for all your testing @Dead2!

Dead2 commented on match_p.h:
```diff
@@ -1,179 +1,44 @@
+#ifdef UNALIGNED_OK
+#if MIN_MATCH >= 4
+typedef uint32_t        bestcmp_t;
+#elif MIN_MATCH > 1
```
`#elif MIN_MATCH > 1` could be `#elif MIN_MATCH >= 2` to match the leading if style

nmoinvaz replied:
Done.

mscastanho commented:
I haven't had the time to review this yet. But I just tested the code on a Power 9 machine, with configure and make, -O2, different files, compression levels and wrappers.

Compression and decompression throughput is pretty much the same over all tests, with some small variance in some cases (between -0.65% and +0.65%), it's hard to distinguish from noise. Results collapsed below:
<details><summary>Test results</summary>
<p>

```
File 'pngpixels'

 - Wrapper 'zlib'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    84.4 |       84.5 |   +0.12% |
   |     2 |    70.8 |       70.8 |   +0.00% |
   |     3 |    43.4 |       43.4 |   +0.00% |
   |     4 |    45.3 |       45.3 |   +0.00% |
   |     5 |    28.1 |       28.1 |   +0.00% |
   |     6 |    14.1 |       14.1 |   +0.00% |
   |     7 |     9.0 |        9.0 |   +0.00% |
   |     8 |     2.8 |        2.8 |   +0.00% |
   |     9 |     1.3 |        1.3 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   259.4 |      259.4 |   +0.00% |
   |     2 |   278.2 |      278.2 |   +0.00% |
   |     3 |   315.0 |      315.0 |   +0.00% |
   |     4 |   240.1 |      240.1 |   +0.00% |
   |     5 |   256.8 |      256.8 |   +0.00% |
   |     6 |   291.1 |      291.1 |   +0.00% |
   |     7 |   305.8 |      305.8 |   +0.00% |
   |     8 |   335.1 |      335.0 |   -0.03% |
   |     9 |   351.1 |      351.1 |   +0.00% |

 - Wrapper 'gzip'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    80.5 |       80.5 |   +0.00% |
   |     2 |    67.9 |       67.9 |   +0.00% |
   |     3 |    42.3 |       42.3 |   +0.00% |
   |     4 |    44.1 |       44.1 |   +0.00% |
   |     5 |    27.7 |       27.7 |   +0.00% |
   |     6 |    14.0 |       14.0 |   +0.00% |
   |     7 |     9.0 |        9.0 |   +0.00% |
   |     8 |     2.8 |        2.8 |   +0.00% |
   |     9 |     1.3 |        1.3 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   225.6 |      225.6 |   +0.00% |
   |     2 |   239.8 |      239.8 |   +0.00% |
   |     3 |   266.5 |      266.5 |   +0.00% |
   |     4 |   210.9 |      210.9 |   +0.00% |
   |     5 |   223.6 |      223.7 |   +0.04% |
   |     6 |   249.2 |      249.2 |   +0.00% |
   |     7 |   260.0 |      259.9 |   -0.04% |
   |     8 |   280.8 |      280.8 |   +0.00% |
   |     9 |   291.9 |      291.9 |   +0.00% |

 - Wrapper 'raw'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    89.7 |       89.7 |   +0.00% |
   |     2 |    74.5 |       74.4 |   -0.13% |
   |     3 |    44.7 |       44.7 |   +0.00% |
   |     4 |    46.8 |       46.7 |   -0.21% |
   |     5 |    28.7 |       28.7 |   +0.00% |
   |     6 |    14.3 |       14.3 |   +0.00% |
   |     7 |     9.1 |        9.1 |   +0.00% |
   |     8 |     2.8 |        2.8 |   +0.00% |
   |     9 |     1.3 |        1.3 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   310.9 |      310.8 |   -0.03% |
   |     2 |   338.2 |      338.2 |   +0.00% |
   |     3 |   394.3 |      394.3 |   +0.00% |
   |     4 |   283.6 |      283.5 |   -0.04% |
   |     5 |   307.1 |      307.0 |   -0.03% |
   |     6 |   357.5 |      357.4 |   -0.03% |
   |     7 |   379.9 |      379.9 |   +0.00% |
   |     8 |   426.1 |      426.2 |   +0.02% |
   |     9 |   452.5 |      452.4 |   -0.02% |

File 'jpeg'

 - Wrapper 'zlib'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    31.1 |       31.3 |   +0.64% |
   |     2 |    30.8 |       30.8 |   +0.00% |
   |     3 |    30.8 |       30.9 |   +0.32% |
   |     4 |    30.8 |       30.8 |   +0.00% |
   |     5 |    30.6 |       30.8 |   +0.65% |
   |     6 |    30.8 |       30.6 |   -0.65% |
   |     7 |    30.8 |       30.8 |   +0.00% |
   |     8 |    30.8 |       30.8 |   +0.00% |
   |     9 |    30.8 |       30.8 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   360.9 |      360.8 |   -0.03% |
   |     2 |   313.6 |      313.9 |   +0.10% |
   |     3 |   322.4 |      322.4 |   +0.00% |
   |     4 |   777.8 |      777.1 |   -0.09% |
   |     5 |   777.6 |      777.7 |   +0.01% |
   |     6 |   777.2 |      777.4 |   +0.03% |
   |     7 |   777.6 |      777.6 |   +0.00% |
   |     8 |   777.2 |      777.7 |   +0.06% |
   |     9 |   777.3 |      777.0 |   -0.04% |

 - Wrapper 'gzip'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    30.6 |       30.7 |   +0.33% |
   |     2 |    30.2 |       30.2 |   +0.00% |
   |     3 |    30.2 |       30.3 |   +0.33% |
   |     4 |    30.2 |       30.2 |   +0.00% |
   |     5 |    30.2 |       30.2 |   +0.00% |
   |     6 |    30.1 |       30.2 |   +0.33% |
   |     7 |    30.2 |       30.2 |   +0.00% |
   |     8 |    30.2 |       30.2 |   +0.00% |
   |     9 |    30.2 |       30.2 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   296.6 |      296.6 |   +0.00% |
   |     2 |   264.1 |      264.1 |   +0.00% |
   |     3 |   270.1 |      270.1 |   +0.00% |
   |     4 |   530.2 |      529.9 |   -0.06% |
   |     5 |   530.2 |      530.2 |   +0.00% |
   |     6 |   530.0 |      530.1 |   +0.02% |
   |     7 |   529.6 |      530.1 |   +0.09% |
   |     8 |   530.1 |      530.1 |   +0.00% |
   |     9 |   529.7 |      530.1 |   +0.08% |

 - Wrapper 'raw'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    31.9 |       31.9 |   +0.00% |
   |     2 |    31.5 |       31.4 |   -0.32% |
   |     3 |    31.5 |       31.4 |   -0.32% |
   |     4 |    31.5 |       31.4 |   -0.32% |
   |     5 |    31.5 |       31.5 |   +0.00% |
   |     6 |    31.5 |       31.5 |   +0.00% |
   |     7 |    31.3 |       31.5 |   +0.64% |
   |     8 |    31.5 |       31.5 |   +0.00% |
   |     9 |    31.5 |       31.5 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   474.7 |      474.6 |   -0.02% |
   |     2 |   396.5 |      396.6 |   +0.03% |
   |     3 |   410.1 |      410.3 |   +0.05% |
   |     4 |  1605.6 |     1606.3 |   +0.04% |
   |     5 |  1605.9 |     1605.5 |   -0.02% |
   |     6 |  1607.5 |     1608.0 |   +0.03% |
   |     7 |  1606.2 |     1607.9 |   +0.11% |
   |     8 |  1606.6 |     1607.6 |   +0.06% |
   |     9 |  1606.8 |     1607.8 |   +0.06% |

File 'executable'

 - Wrapper 'zlib'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    54.4 |       54.4 |   +0.00% |
   |     2 |    48.6 |       48.6 |   +0.00% |
   |     3 |    35.0 |       35.0 |   +0.00% |
   |     4 |    33.8 |       33.8 |   +0.00% |
   |     5 |    23.0 |       23.0 |   +0.00% |
   |     6 |    13.6 |       13.6 |   +0.00% |
   |     7 |    10.1 |       10.1 |   +0.00% |
   |     8 |     5.6 |        5.6 |   +0.00% |
   |     9 |     4.2 |        4.2 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   175.4 |      175.4 |   +0.00% |
   |     2 |   181.1 |      181.1 |   +0.00% |
   |     3 |   187.6 |      187.7 |   +0.05% |
   |     4 |   179.7 |      179.7 |   +0.00% |
   |     5 |   183.9 |      183.9 |   +0.00% |
   |     6 |   189.1 |      189.1 |   +0.00% |
   |     7 |   190.5 |      190.6 |   +0.05% |
   |     8 |   193.2 |      193.2 |   +0.00% |
   |     9 |   193.4 |      193.4 |   +0.00% |

 - Wrapper 'gzip'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    52.7 |       52.7 |   +0.00% |
   |     2 |    47.3 |       47.2 |   -0.21% |
   |     3 |    34.3 |       34.3 |   +0.00% |
   |     4 |    33.1 |       33.1 |   +0.00% |
   |     5 |    22.6 |       22.6 |   +0.00% |
   |     6 |    13.5 |       13.5 |   +0.00% |
   |     7 |    10.0 |       10.0 |   +0.00% |
   |     8 |     5.5 |        5.5 |   +0.00% |
   |     9 |     4.2 |        4.2 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   159.5 |      159.5 |   +0.00% |
   |     2 |   164.2 |      164.2 |   +0.00% |
   |     3 |   169.6 |      169.6 |   +0.00% |
   |     4 |   163.1 |      163.1 |   +0.00% |
   |     5 |   166.5 |      166.5 |   +0.00% |
   |     6 |   170.7 |      170.7 |   +0.00% |
   |     7 |   171.9 |      171.9 |   +0.00% |
   |     8 |   174.1 |      174.1 |   +0.00% |
   |     9 |   174.3 |      174.2 |   -0.06% |

 - Wrapper 'raw'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    56.5 |       56.4 |   -0.18% |
   |     2 |    50.3 |       50.3 |   +0.00% |
   |     3 |    35.8 |       35.9 |   +0.28% |
   |     4 |    34.5 |       34.5 |   +0.00% |
   |     5 |    23.3 |       23.3 |   +0.00% |
   |     6 |    13.7 |       13.7 |   +0.00% |
   |     7 |    10.1 |       10.1 |   +0.00% |
   |     8 |     5.6 |        5.6 |   +0.00% |
   |     9 |     4.2 |        4.2 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   197.1 |      197.1 |   +0.00% |
   |     2 |   204.4 |      204.4 |   +0.00% |
   |     3 |   212.8 |      212.7 |   -0.05% |
   |     4 |   202.6 |      202.6 |   +0.00% |
   |     5 |   208.0 |      208.0 |   +0.00% |
   |     6 |   214.6 |      214.6 |   +0.00% |
   |     7 |   216.5 |      216.5 |   +0.00% |
   |     8 |   219.9 |      219.9 |   +0.00% |
   |     9 |   220.1 |      220.2 |   +0.05% |

File 'html'

 - Wrapper 'zlib'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    54.6 |       54.5 |   -0.18% |
   |     2 |    47.5 |       47.5 |   +0.00% |
   |     3 |    32.5 |       32.5 |   +0.00% |
   |     4 |    31.9 |       31.9 |   +0.00% |
   |     5 |    19.7 |       19.7 |   +0.00% |
   |     6 |    12.9 |       12.9 |   +0.00% |
   |     7 |    11.1 |       11.1 |   +0.00% |
   |     8 |     8.3 |        8.3 |   +0.00% |
   |     9 |     8.2 |        8.2 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   197.7 |      197.7 |   +0.00% |
   |     2 |   204.6 |      204.6 |   +0.00% |
   |     3 |   213.1 |      213.3 |   +0.09% |
   |     4 |   205.5 |      205.5 |   +0.00% |
   |     5 |   204.3 |      204.3 |   +0.00% |
   |     6 |   208.5 |      208.6 |   +0.05% |
   |     7 |   208.8 |      208.8 |   +0.00% |
   |     8 |   209.2 |      209.2 |   +0.00% |
   |     9 |   209.0 |      209.0 |   +0.00% |

 - Wrapper 'gzip'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    52.9 |       52.9 |   +0.00% |
   |     2 |    46.3 |       46.3 |   +0.00% |
   |     3 |    31.9 |       31.9 |   +0.00% |
   |     4 |    31.4 |       31.4 |   +0.00% |
   |     5 |    19.5 |       19.5 |   +0.00% |
   |     6 |    12.8 |       12.8 |   +0.00% |
   |     7 |    11.0 |       11.0 |   +0.00% |
   |     8 |     8.3 |        8.3 |   +0.00% |
   |     9 |     8.2 |        8.2 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   177.9 |      178.0 |   +0.06% |
   |     2 |   183.5 |      183.5 |   +0.00% |
   |     3 |   190.5 |      190.5 |   +0.00% |
   |     4 |   184.3 |      184.3 |   +0.00% |
   |     5 |   183.3 |      183.3 |   +0.00% |
   |     6 |   186.7 |      186.7 |   +0.00% |
   |     7 |   186.8 |      186.9 |   +0.05% |
   |     8 |   187.1 |      187.1 |   +0.00% |
   |     9 |   187.1 |      187.1 |   +0.00% |

 - Wrapper 'raw'

  -> comp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |    56.5 |       56.5 |   +0.00% |
   |     2 |    49.0 |       49.0 |   +0.00% |
   |     3 |    33.2 |       33.2 |   +0.00% |
   |     4 |    32.6 |       32.6 |   +0.00% |
   |     5 |    20.0 |       20.0 |   +0.00% |
   |     6 |    13.0 |       13.0 |   +0.00% |
   |     7 |    11.2 |       11.2 |   +0.00% |
   |     8 |     8.4 |        8.4 |   +0.00% |
   |     9 |     8.3 |        8.3 |   +0.00% |

  -> uncomp_throughput_MB
   | level | default | compare258 |     gain |
   |     1 |   225.4 |      225.4 |   +0.00% |
   |     2 |   234.3 |      234.3 |   +0.00% |
   |     3 |   245.8 |      245.8 |   +0.00% |
   |     4 |   235.6 |      235.6 |   +0.00% |
   |     5 |   233.9 |      233.9 |   +0.00% |
   |     6 |   239.6 |      239.7 |   +0.04% |
   |     7 |   239.8 |      239.8 |   +0.00% |
   |     8 |   240.4 |      240.3 |   -0.04% |
   |     9 |   240.2 |      240.1 |   -0.04% |
```
</p>
</details>

nmoinvaz commented:
I have changed `#elif MIN_MATCH > 1` to `#elif MIN_MATCH >= 2` as suggested.

nmoinvaz commented:
I missed two instances of that byte check in compare258 for 32bit and 64bit.

Dead2 commented:
```
# PR/550/10 9b6615d8108bd3c410f94972e93fd63ae2a245c9
   text    data     bss     dec     hex filename
  90461    1280      32   91773   1667d libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.223/0.229/0.231       0.128/0.131/0.134  20804915
 2      38.062%    0.431/0.437/0.441       0.131/0.135/0.137  15964205
 3      37.021%    0.525/0.538/0.545       0.125/0.130/0.132  15527698
 4      35.832%    0.594/0.600/0.605       0.122/0.128/0.131  15028898
 5      35.719%    0.644/0.656/0.661       0.124/0.128/0.130  14981562
 6      35.651%    0.775/0.792/0.802       0.124/0.130/0.132  14953264
 7      35.030%    1.447/1.462/1.472       0.123/0.129/0.132  14692564
 8      34.995%    2.233/2.244/2.256       0.124/0.127/0.130  14677992

 avg1   37.739%                0.870                   0.130
 tot                         104.370                  15.563
```

nmoinvaz committed:
Unroll more in compare258_c for performance improvement.
Unify length count variable across all compare256 variants.
Early return without break for possible performance improvements.

nmoinvaz commented:
I moved the #undefs to _match_p.h_ so I don't have to undef multiple times in compare258.c for each `LONGEST_MATCH`, `COMPARE256`, ..

nmoinvaz commented:
I have added another commit after I did some _compare258_ function testing.
https://gist.github.com/nmoinvaz/d4b4f71237a2dd877d2efe124d72c241
It fixes length calculation issues with `compare258_c` and `compare258_unaligned_16`.

It also adds some possible performance improvements (release mode) to the compare256 variants.

![image](https://user-images.githubusercontent.com/1364432/82720467-b8d94500-9c68-11ea-97d5-601618be9810.png)

The `_x2` versions are after my last commit. And the original are before the commit.
https://gist.github.com/nmoinvaz/1bf9e4098fe6399a14f829957c2123f8

The performance isn't as great in debug mode (obviously). PowerPC @mscastanho might see some improvement at least due to the extra unrolling in compare256_c. Any more unrolling than that for _c_ variant and the performance gains are nil.

nmoinvaz commented:
Cleaned up my last commit a bit.

mtl1979 commented:
CI is failing on multiple arches (PPC, PPC64, PPC64LE, SPARC64 and S390X)

nmoinvaz commented:
@mtl1979  Thanks, sometimes I forget to come back and check on these things.

nmoinvaz commented:
Ok I think I am done on this one for now. Perhaps it needs to be performance tested another time. @Dead2

Dead2 commented:
```
# PR/550/11 dd504a67cc58b6126aad3fbf76e9674d51b4056a
   text    data     bss     dec     hex filename
  90813    1280      32   92125   167dd libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max
 1      49.603%    0.226/0.229/0.231       0.129/0.133/0.135  20804915
 2      38.062%    0.432/0.438/0.441       0.130/0.136/0.138  15964205
 3      37.021%    0.529/0.536/0.543       0.129/0.134/0.135  15527698
 4      35.832%    0.586/0.597/0.602       0.126/0.129/0.131  15028898
 5      35.719%    0.638/0.647/0.653       0.126/0.129/0.132  14981562
 6      35.651%    0.776/0.796/0.804       0.127/0.130/0.131  14953264
 7      35.030%    1.447/1.462/1.477       0.120/0.127/0.130  14692564
 8      34.995%    2.232/2.251/2.270       0.125/0.129/0.131  14677992

 avg1   37.739%                0.870                   0.131
 tot                         104.352                  15.696
```

Looking good. Are there more changes, or is this ready to stabilize for merge?

nmoinvaz commented:
No more changes on my end.
